src/search/landmarks/landmark_factory.cc

```diff
@@ -148,35 +153,44 @@ void LandmarkFactory::mk_acyclic_graph() {
                  << " reasonable or obedient reasonable orders" << endl;
 }
 
-bool LandmarkFactory::remove_first_weakest_cycle_edge(LandmarkNode *cur,
-                                                      list<pair<LandmarkNode *, EdgeType>> &path,
-                                                      list<pair<LandmarkNode *, EdgeType>>::iterator it) {
-    LandmarkNode *parent_p = 0;
-    LandmarkNode *child_p = 0;
-    for (list<pair<LandmarkNode *, EdgeType>>::iterator it2 = it; it2
-         != path.end(); ++it2) {
+void LandmarkFactory::remove_first_weakest_cycle_edge(
+    list<pair<LandmarkNode *, EdgeType>> &path,
+    list<pair<LandmarkNode *, EdgeType>>::iterator it) {
+    LandmarkNode *from = path.back().first;
+    LandmarkNode *to = it->first;
+    EdgeType weakest_edge = path.back().second;
+    for (list<pair<LandmarkNode *, EdgeType>>::iterator it2 = it;
+         it2 != prev(path.end()); ++it2) {
         EdgeType edge = it2->second;
-        if (edge == EdgeType::REASONABLE || edge == EdgeType::OBEDIENT_REASONABLE) {
-            parent_p = it2->first;
-            if (*it2 == path.back()) {
-                child_p = cur;
-                break;
-            } else {
-                list<pair<LandmarkNode *, EdgeType>>::iterator child_it = it2;
-                ++child_it;
-                child_p = child_it->first;
-            }
-            if (edge == EdgeType::OBEDIENT_REASONABLE)
-                break;
-            // else no break since o_r order could still appear in list
+        if (edge < weakest_edge) {
+            from = it2->first;
+            to = next(it2)->first;
+            weakest_edge = edge;
+        }
+        if (weakest_edge == EdgeType::OBEDIENT_REASONABLE) {
+            break;
+        }
+    }
+    if (weakest_edge > EdgeType::REASONABLE) {
+        /*
+          If the weakest ordering is natural (or stronger), this means
+          the entire cycle consists of only natural (or stronger)
+          orderings which indicates that the task is unsolvable. We
+          signal this by clearing the *first_achievers* of all landmarks
+          in the cycle.
+          Note that *calc_achievers* should never be called after
+          *mk_acyclic_graph* because of this, because otherwise this
+          information is lost.
+        */
+        for (list<pair<LandmarkNode *, EdgeType>>::iterator it2 = it;
```

Comment:
This could probably (and hence) should be a range-based for loop, I think. I guess it's the way it is because it was that way before, just a note for similar things in the future. (In the diff it shows up as "new code", but I guess that's because the diff couldn't match the old to the new perfectly here.)